wp-content/plugins/tapestry/controller/class.user-controller.php

```diff
@@ -0,0 +1,140 @@
+<?php
+// TODO Change exceptions to using an ERROR class
+/**
+ * Add/update/retrieve User progress
+ * 
+ */
+class TapestryUserController {
+
+    private $userId = NULL;
+
+    public function __construct() {
+        $this->userId = apply_filters('determine_current_user', false); 
+    }
+
+    /** 
+     * Update User's video progress for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @nodeId the current node being viewed
+     * @param type @progressValue is how much the video was viewed, value should be between >= 0 and <= 1
+    */
+    public function updateProgress($postId, $nodeId, $progressValue) {
+        $this->_checkUserAndPostId($postId);
+
+        if ($progressValue !== NULL) {
+            $progressValue = floatval($progressValue);
+        }
+
+        // Value should be between 0 and 1
+        if ($progressValue < 0 || $progressValue > 1) {
+            throw new Exception('Invalid progress value');
+        }
+
+        $this->_updateUserProgress($postId, $nodeId, $progressValue);
+    }
+
+    /** 
+     * Get User's video progress for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @nodeIdArr is a list of ids currently in the tapestry
+    */
+    public function getProgress($postId, $nodeIdArr) {
+        $this->_isValidTapestryPost($postId);
+        $this->_checkUserAndPostId($postId);
+
+        return $this->_getUserProgress($postId, $nodeIdArr);
+    }
+
+    /** 
+     * Update User's h5p video setting for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @h5pSettingsData stores volume, playbackRate, quality of h5p video
+    */
+    public function updateH5PSettings($postId, $h5pSettingsData) {
+        $this->_checkUserAndPostId($postId);
+
+        if ($this->_isJson($h5pSettingsData)) {
+            $h5pSettingsData = json_decode($h5pSettingsData);
+        } else {
+            throw new Exception('Invalid json');
+        }
+
+        $this->_updateUserH5PSettings($postId, $h5pSettingsData);
+    }
+
+    /** 
+     * Get User's h5p video setting for a tapestry post
+     * @param type @postId the post's Id
+    */
+    public function getH5PSettings($postId) {
+        $this->_isValidTapestryPost($postId);
+        $this->_checkUserAndPostId($postId);
+        return $this->_getUserH5PSettings($postId);
+    }
+
+    private function _updateUserProgress($postId, $nodeId, $progressValue) {
+        $userId = $this->userId;
```

Comment:
I would just use `$this->userId` directly in the function below

Reply:
Can't seem to reply to the getTaptestryNodeIds comment. One where it says "Please use the getTapestryNodeIds function you added in the tapestry controller.". It is using it in the endpoint.